src/esp/gfx/ssao/HBAOHelper.cpp

```diff
@@ -0,0 +1,724 @@
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glBindTexture(GL_TEXTURE_2D, 0);
+  }
+
+  nvgl::newTexture(textures.hbao2_resultarray, GL_TEXTURE_2D_ARRAY);
+  glBindTexture(GL_TEXTURE_2D_ARRAY, textures.hbao2_resultarray);
+  glTexStorage3D(GL_TEXTURE_2D_ARRAY, 1, formatAO, quarterWidth, quarterHeight,
+                 HBAO_RANDOM_ELEMENTS);
+  glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+  glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+  glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+  glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+  glBindTexture(GL_TEXTURE_2D_ARRAY, 0);
+
+  GLenum drawbuffers[NUM_MRT];
+  for (int layer = 0; layer < NUM_MRT; layer++) {
+    drawbuffers[layer] = GL_COLOR_ATTACHMENT0 + layer;
+  }
+
+  nvgl::newFramebuffer(fbos.hbao2_deinterleave);
+  glBindFramebuffer(GL_FRAMEBUFFER, fbos.hbao2_deinterleave);
+  glDrawBuffers(NUM_MRT, drawbuffers);
+  glBindFramebuffer(GL_FRAMEBUFFER, 0);
+
+  nvgl::newFramebuffer(fbos.hbao2_calc);
+  glBindFramebuffer(GL_FRAMEBUFFER, fbos.hbao2_calc);
+#if USE_AO_LAYERED_SINGLEPASS == AO_LAYERED_IMAGE
+  // this fbo will not have any attachments and therefore requires rasterizer to
+  // be configured through default parameters
+  glFramebufferParameteri(GL_FRAMEBUFFER, GL_FRAMEBUFFER_DEFAULT_WIDTH,
+                          quarterWidth);
+  glFramebufferParameteri(GL_FRAMEBUFFER, GL_FRAMEBUFFER_DEFAULT_HEIGHT,
+                          quarterHeight);
+#endif
+#if USE_AO_LAYERED_SINGLEPASS == AO_LAYERED_GS
+  glFramebufferTexture(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0,
+                       textures.hbao2_resultarray, 0);
+#endif
+  glBindFramebuffer(GL_FRAMEBUFFER, 0);
+
+  return true;
+}
+
+void HBAOHelper::prepareHbaoData(const Projection& projection,
+                                 int width,
+                                 int height) {
+  // projection
+  const float* P = projection.matrix.get_value();
+
+  float projInfoPerspective[] = {
+      2.0f / (P[4 * 0 + 0]),                  // (x) * (R - L)/N
+      2.0f / (P[4 * 1 + 1]),                  // (y) * (T - B)/N
+      -(1.0f - P[4 * 2 + 0]) / P[4 * 0 + 0],  // L/N
+      -(1.0f + P[4 * 2 + 1]) / P[4 * 1 + 1],  // B/N
+  };
+
+  float projInfoOrtho[] = {
+      2.0f / (P[4 * 0 + 0]),                  // ((x) * R - L)
+      2.0f / (P[4 * 1 + 1]),                  // ((y) * T - B)
+      -(1.0f + P[4 * 3 + 0]) / P[4 * 0 + 0],  // L
+      -(1.0f - P[4 * 3 + 1]) / P[4 * 1 + 1],  // B
+  };
+
+  int useOrtho = projection.ortho ? 1 : 0;
+  m_hbaoUbo.projOrtho = useOrtho;
+  m_hbaoUbo.projInfo = useOrtho ? projInfoOrtho : projInfoPerspective;
+
+  float projScale;
+  if (useOrtho) {
+    projScale = float(height) / (projInfoOrtho[1]);
+  } else {
+    projScale = float(height) / (tanf(projection.fov * 0.5f) * 2.0f);
```

Comment:
Note : this is incorrect - tanf works on radians, not degrees - and subsequently the effect seen is much stronger than it should be.